### PR TITLE
Support remote download of files when doing a File.Get

### PR DESCRIPTION
**Pull request by ejbrever**
@aashaikh @robshakir Can you please review this?

### CODE REVIEW

File: common.proto [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+syntax = "proto3";
+
+package gnoi;
+
+import "third_party/openconfig/gnoi/types.proto";
```
Comment (aashaikh) [inline common.proto]: import path cleanup

File: file/file.proto [hunk 2]
```diff
@@ -28,6 +29,13 @@ service File {
   // if the file does not exist or there was an error reading the file.
```
Comment (aashaikh) [inline file/file.proto]: "Get reads and streams ..."

File: system/system.proto [hunk 3]
```diff
@@ -305,7 +285,7 @@ message Package {
   // For system packages this will take effect after reboot.
   bool activate = 5;
```
Comment (aashaikh) [inline system/system.proto]: I still wonder if `activate` is the right name for this field -- to me a true value here could be interpreted as "make this package active now".  does something like `requires_reboot` make sense ?

File: common.proto (hunk 1 above)
Comment (ejbrever) [reply common.proto]: oops, thanks.

File: file/file.proto [hunk 4]
```diff
@@ -18,6 +18,7 @@ syntax = "proto3";
 
 package gnoi.file;
 
+import "third_party/openconfig/gnoi/common.proto";
```
[the remark this answers is not in the document]
Comment (ejbrever) [reply file/file.proto]: thanks :)

File: file/file.proto (hunk 2 above)
Comment (ejbrever) [reply file/file.proto]: done

File: system/system.proto [hunk 5]
```diff
@@ -24,6 +24,7 @@ syntax = "proto3";
 
 package gnoi.system;
 
+import "third_party/openconfig/gnoi/common.proto";
```
[the remark this answers is not in the document]
Comment (ejbrever) [reply system/system.proto]: Fixed

File: system/system.proto (hunk 3 above)
Comment (ejbrever) [reply system/system.proto]: I am open to a new name, just not sure what that might be :)

For more context, when you transfer a file, the file can be stored on the device, but not applied. So after transferring a file (without activation) a reboot of the device would not set the package to active. You have to specify somehow to set a specific package active.

File: file/file.proto [hunk 6]
```diff
@@ -96,6 +104,21 @@ message GetResponse {
   }
 }
 
+// A GetRemoteRequest specifies the remote_file to transfer and the details on
+// where to transfer the data. The remote_file must be an absolute path to a
+// file.
+message GetRemoteRequest {
+  string remote_file = 1;
```
[the remark this answers is not in the document]
Comment (ejbrever) [reply file/file.proto]: Yea, agree this is a bit confusing and the docs don't help :)

The intention was RemoteDownload specified how to get the file, and this was supposed to specify the path name in the device to save it to (i.e. /tmp/os-1.0).

Maybe local_file be a better name and I can update the descriptions as well. Please let me know what you think.

[the remark this answers is not in the document]
Comment (robshakir) [reply file/file.proto]: Yep - something like `local_path` would probably make more sense here IMHO.

File: system/system.proto (hunk 3 above)
Comment (aashaikh) [reply system/system.proto]: I think I'm getting hung up on the "reboot" part of this -- maybe we can change the comment to say something like:
```
Indicates that the package should be made active after receipt on the device.
For system image packages, the new image is expected to be active after a reboot.
```

File: common.proto [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+syntax = "proto3";
+
+package gnoi;
+
+import "github.com/openconfig/gnoi/types.proto";
+
+// RemoteDownload defines the details for a device to initiate a file transfer
+// from or to a remote location.
+message RemoteDownload {
```
Comment (aghaffarkhah) [inline common.proto]: Sorry if I missed this, but why do you think this message is common to all the services? What is the reason behind creating this file? Sorry this is just me understanding...

File: file/file.proto [hunk 8]
```diff
@@ -18,16 +18,24 @@ syntax = "proto3";
...
   // that contains the hash of the data sent. An error is returned
   // if the file does not exist or there was an error reading the file.
   rpc Get(GetRequest) returns (stream GetResponse) {}
 
+  // GetRemote transfers the contents of a file from the target to a specified
+  // remote location. The response contains the hash of the data transferred.
+  // An error is returned if the file does not exist, the file transfer fails,
+  // or if there was an error reading the file. This is a blocking call until
+  // the file transfer is complete.
+  rpc GetRemote(GetRemoteRequest) returns (GetRemoteResponse) {}
```
Comment (aghaffarkhah) [inline file/file.proto]: Wouldnt "TransferToRemote" or something similar a better name? Sorry not trying to be obsessed with the name, but having "Get" in the name of the RPC is a bit confusing.

File: common.proto (hunk 7 above)
Comment (ejbrever) [reply common.proto]: No problem at all. It was originally in system.proto. In this change we wanted to add it to file.proto as well so it made sense to put it in a common place.

File: file/file.proto (hunk 8 above)
Comment (ejbrever) [reply file/file.proto]: I'm pretty open to a different name. Maybe TransferFromRemote would be more obvious here?

@aashaikh @robshakir Any thoughts on the name here as well? I know we had discussed this previously as well.

File: system/system.proto (hunk 3 above)
Comment (ejbrever) [reply system/system.proto]: Sure, done.

File: file/file.proto (hunk 6 above)
[the remark this answers is not in the document]
Comment (ejbrever) [reply file/file.proto]: Sure, done.

File: file/file.proto (hunk 8 above)
Comment (robshakir) [reply file/file.proto]: I'm fine with `TransferFromRemote` or even `FetchRemote` here, but I don't have strong feelings. 

The "get" in `GetRemote` is essentially "get this file from a remote place" - maybe fetch is better than "get" here?

Comment (ejbrever) [reply file/file.proto]: I'm leaning towards TransferFromRemote, but please let me know if you feel otherwise.

Comment (ejbrever) [reply file/file.proto]: After thinking again about it, I agree with @aghaffarkhah on TransferToRemote :)

I have updated using this name now. I think this is much more obvious than GetRemote.

**Comment by ejbrever** [comment]
@robshakir @aghaffarkhah Can you please take another look?

Thanks,
Eric

**Pull request merged**